.github/actions/determine-or-use-target-branch-and-get-last-released-image/action.yaml

```diff
@@ -0,0 +1,103 @@
+name: "Determine or Use Target Branch and Get Latest Docker Image"
+description: |
+  Determine the branch to fetch the latest docker image tag from, or use a target branch directly.
+  - If base-branch is set, determine the target branch to fetch the latest docker image tag from
+    - Determine the IMAGE_TAG based on the latest release R and check for images built on branch R - 1:
+      - For commits on the aptos-release-v1.19 branch, the IMAGE_TAG should be the latest commit built on aptos-release-v1.18
+      - For commits on the main branch, the IMAGE_TAG should be the latest commit on the max release aptos-release-v<X.X> branch
+      - For commits on other branches, the IMAGE_TAG should be the latest commit on the branch's last release aptos-release-v<X.X> branch
+  - If branch is set, use it directly
+
+inputs:
+  base-branch:
+    description: "The base branch to determine the target from, or use a target branch directly"
+    required: false
+  branch:
+    description: "Use this branch directly if set"
+    required: false
+  if-need-to-determine-target-branch:
+    description: "Set to true if need to determine target branch based on base branch"
+    required: false
+    default: "false"
+  variants:
+    description: "The variants to check, as a space-separated string, e.g. 'performance failpoints'"
+    required: false
+
+outputs:
+  TARGET_BRANCH:
+    description: "The determined or target target branch"
+    value: ${{ steps.set-target-branch.outputs.TARGET_BRANCH }}
+  IMAGE_TAG:
+    description: "The latest docker image tag for the given branch and variants"
+    value: ${{ steps.determine-test-image-tag.outputs.IMAGE_TAG }}
+
+runs:
+  using: composite
+  steps:
+    # Checkout repository based on base branch or target branch
+    - name: Checkout base branch
+      if: ${{ inputs.if-need-to-determine-target-branch == 'true' }}
+      uses: actions/checkout@v4
+      with:
+        ref: ${{ inputs.base-branch }}
+        path: checkout_branch
+        fetch-depth: 0
+
+    # If using the target branch directly
+    - name: Checkout target branch
+      if: ${{ inputs.if-need-to-determine-target-branch == 'false' }}
+      uses: actions/checkout@v3
+      with:
+        ref: ${{ inputs.branch }}
+        path: checkout_branch
+        fetch-depth: 0
+
+    - name: Setup Python environment
```

Comment:
Interesting here we really see the shortcoming of using python for github actions

Currently we do not have a good model for github action distribution and versioning (using python actions)

In typescript, we actually can build the action and tag and release them

We could in theory build a python action distribution mechanism whereby we build a binary and then in the preamble for the action download and unpack that binary and run from the composite action

Then you dont need any of this double checking out stuff / python setup

However, that is a problem for future us :)

This is fine as is, just wanna share my thoughts on this issue